### ISSUE TITLE
Adaptive LR scheduling for non-stationary training distributions

### ISSUE THREAD

**Issue by cweniger** [cweniger/falcon#81]
## Background

Falcon's training loop uses `ReduceLROnPlateau` with a fixed patience (currently 16 epochs). This is problematic because the training distribution evolves continuously as the proposal buffer is refreshed with new samples. The scheduler cannot distinguish between:

- **Genuine plateau**: model has converged on the current distribution → decay LR
- **Distribution shift noise**: val_loss oscillates because new buffer samples arrived → don't decay

This causes a cascade failure mode: premature LR reductions → model can't track evolving buffer → more oscillation → more reductions → runaway decay.

## Observed dynamics

From profiling the 04_gaussian example, the buffer refreshes every ~23 epochs (128 new samples added at a time into a 768-sample training cache). A full buffer replacement takes ~138 epochs (~1100 gradient steps). The loss shows a characteristic staircase pattern at this timescale — each step is a phase transition as the buffer completes a refresh cycle, not an optimizer event.

## Key design constraints

- Training duration and network complexity are not known in advance
- Must work for simple MLPs and complex CNN embeddings equally
- Simulation speed vs training speed ratio is unpredictable
- Standard in related literature (RL, sequential SBI) is **constant LR with Adam**

### 1. Constant LR (simplest, RL standard)
Just remove the scheduler. Adam's internal `v_t` accumulation already provides implicit per-parameter decay. The staircase dynamics show this works well in practice.

### 2. LR finder at initialization (Smith 2017)
Run a short pre-training phase (100–200 steps) that ramps LR exponentially from 1e-6 to 1e-1, records the loss at each step, then picks the LR just below the divergence point (steepest descent on the loss-vs-LR curve).

Benefits:
- Architecture-agnostic: works for identity embeddings, deep CNNs, or anything else
- No manual tuning: the right LR is derived from the actual network and initial data
- Cheap: ~100 steps before the main training loop, negligible cost
- Should be run on the **initial buffer** (prior samples) since that is the hardest distribution the model will ever see

Config addition to `OptimizerConfig`:
- `lr_finder_steps: int = 0` — if > 0, run the ramp before training; 0 = disabled
- `lr_finder_min: float = 1e-6`
- `lr_finder_max: float = 1e-1`
- `lr_finder_factor: float = 0.1` — pick LR this factor below the divergence point

The found LR then becomes the constant LR (or the ceiling for any decay schedule).

### 3. Stability flag + conditional scheduling
Maintain an internal `_is_stable` flag based on the rate of change of `_output_std` (already computed in `GaussianPosterior`). The LR scheduler only acts during stable phases; LR resets to initial value when instability is detected (buffer shift).

```python
output_std_change = abs(current_std - prev_std) / prev_std
_is_stable = (output_std_change < stability_threshold) for N consecutive epochs
if _is_stable:
    scheduler.step(val_loss)  # or fixed-rate decay
else:
    reset lr to initial value
```

Config additions to `OptimizerConfig`:
- `stability_threshold: float` — max relative change in `_output_std` per epoch
- `stability_min_epochs: int` — epochs of stability before flag sets
- `lr_decay_mode: str` — `"none"` / `"plateau"` / `"fixed_rate"`
- `lr_decay_rate: float` — per stable epoch (for fixed_rate mode)

### 4. Adaptive β₂ switching
`β₂=0.9` (current) gives a 10-step forgetting time constant — good for tracking distribution shifts but noisy during stable convergence. Standard Adam uses `β₂=0.999` (~1000 steps). Switch between them on the stability flag:

```python
betas = (β₁, 0.9)   if not _is_stable   # fast adaptation
betas = (β₁, 0.999) if _is_stable       # smooth convergence
```

This is more principled than LR scheduling — it controls how much gradient history Adam trusts, which is the right knob for non-stationary settings.

### 5. Optimizer state reset on buffer change
Reset Adam's `m_t` and `v_t` when a major distribution shift is detected. Stale second-moment estimates from the old distribution make Adam sluggish when the landscape changes. Related to continual learning literature on Adam resets.

## Related literature

- **RL** (PPO, SAC, TD3): constant LR with Adam is universal standard
- **Sequential SBI** (SNPE-C/APT, sbi library): constant LR per round
- **Loss of plasticity** (Lyle, Nikishin et al.): warns against LR decay in continual settings
- **Gradient noise scale** (McCandlish et al. 2018): connects batch size to optimal LR
- **LR finder / CLR** (Smith 2017): exponential ramp to find optimal LR before training
- **SGDR** (Smith 2017): warm restarts — buffer-triggered reset is the data-driven version
- **AdaBelief, AMSGrad**: touch on β₂ forgetting rate problem

## Suggested next steps

- Validate constant LR as default (already empirically supported by run_full_cov4)
- Implement LR finder as optional pre-training phase (`lr_finder_steps > 0`)
- Implement `_is_stable` flag as internal estimator state (useful beyond scheduling)
- Prototype β₂ switching — low implementation cost, theoretically motivated
